feat(alerts): Add retry mechanism to percolator content preparation

ERosendo opened the pull request:
This PR adds the retry decorator to the helper function responsible for preparing percolator content (`prepare_percolator_content`). This enhances the robustness of alert generation by automatically retrying operations that might fail due to race conditions, adding a maximum of 2 seconds to affected tasks.

Implements the suggestion from #4623

mlissner commented:
Cool. A few thoughts:

1. If I understand this correctly (I'm not sure I do), this means we're going to be doing a retry nearly every time we create an alert, right? That's a bummer.

2. We don't have heuristics about how many retries we need to do or how long they'll take, but if this is about race conditions, I guess I'd expect it to be resolved in under a second?

3. Since I think we agreed that this solution is pretty bad (but necessary!), shall we create one more issue to refactor it? Maybe we can stub it out during this sprint, analyze it properly during `H`, and then try to fix it in `G`?

ERosendo commented:
> this means we're going to be doing a retry nearly every time we create an alert, right? That's a bummer.

This error doesn't seem to happen very often nowadays. I checked Sentry yesterday and found it occurred just 82 times over the past 30 days. We'll use the retry mechanism/delay only when the error happens.

> but if this is about race conditions, I guess I'd expect it to be resolved in under a second?

That's right, that's why I set the decorator to retry only three times with a maximum delay of one second on the last attempt.

> shall we create one more issue to refactor it? Maybe we can stub it out during this sprint, analyze it properly during H, and then try to fix it in G?

That sounds good 👍

albertisfu commented on cl/alerts/utils.py:
```diff
@@ -619,6 +621,7 @@ def select_es_document_fields(
     }
 
 
+@retry(ObjectDoesNotExist, tries=3, delay=0.25, backoff=2)
```
Considering the issue seems related to the use of async views, database transactions, and connection visibility. When switching from sync to async methods, they use different threads and database connections, so transaction integrity isn’t guaranteed. 

I believe what’s happening is that the `on_commit` hook is triggered before the entire RD transaction is fully committed. Then, the Celery worker picks it up using a DB connection that still holds an old snapshot and doesn’t yet have visibility of the new RD.

If we only retry the method inside the task, the retries will likely use the same DB connection (since they’re in the same thread), and the problem will persist. 

That’s why I think it’s better to retry the whole task,  this way, the retry has a chance to use a different DB connection, which probably does see the newly committed document.

ERosendo replied:
Thanks Alberto! I'll update the code

ERosendo commented:
@albertisfu Thanks for your feedback. 

You're right – retrying the whole task instead of just the helper function is the way to go to ensure we're using an updated DB snapshot. I've updated the approach to use a custom retry for the Celery task and added a setting named `PERCOLATOR_MISSING_DOCUMENT_MAX_RETRIES`. This setting defaults to 4, allowing us to easily adjust it in the future if our initial estimate isn't quite right.

mlissner commented:
I set for auto-merge! Thanks guys.

albertisfu commented:
> ## Suspect Issues
> This pull request was deployed and Sentry observed the following issues:
> 
> * ‼️ **RECAPDocument.DoesNotExist: RECAPDocument matching query does not exist.** `cl.alerts.tasks.send_or_schedule_search_alerts` [View Issue](https://freelawproject.sentry.io/issues/6676054096/?referrer=github-pr-bot)
> 
> Did you find this useful? React with a 👍 or 👎

This is OK. It's related to documents that indeed do not exist.
`442530570 `
`442530585`
